### PR TITLE
[41310] Avoid outputting internal statement errors

### PR DESCRIPTION
**Pull request by oliverguenther**
https://community.openproject.org/wp/41310

### CODE REVIEW

File: lib/api/errors/internal_error.rb [hunk 1]
```diff
@@ -34,15 +34,25 @@ class InternalError < ErrorBase
...
 
         super error
       end
+
+      private
+
+      ##
+      # Hide internal database errors in production
+      def hidden_exception?(exception)
+        return false unless Rails.env.production?
```
Comment (cbliard) [inline lib/api/errors/internal_error.rb:52]:
```suggestion
        Rails.env.production? && exception.is_a?(ActiveRecord::StatementInvalid)
```
I would remove the condition, as it reads better (for me) without `unless`.

File: lib/api/errors/internal_error.rb [hunk 2]
```diff
@@ -34,15 +34,25 @@ class InternalError < ErrorBase
...
+
+        exception.is_a?(ActiveRecord::StatementInvalid)
```
Comment (cbliard) [inline lib/api/errors/internal_error.rb]:
```suggestion
        Rails.env.production? && exception.is_a?(ActiveRecord::StatementInvalid)
```
I would remove the condition, as it reads better (for me) without `unless`.

File: lib/api/errors/internal_error.rb (hunk 1 above)
Comment (ulferts) [inline lib/api/errors/internal_error.rb]: This somewhat duplicates the condition on https://github.com/opf/openproject/blob/dev/lib/api/root_api.rb#L237 which admittedly is on the caller's end.

File: lib/api/errors/internal_error.rb [hunk 3]
```diff
@@ -34,15 +34,25 @@ class InternalError < ErrorBase
       identifier 'InternalServerError'
       code 500
 
-      def initialize(error_message = nil, **)
+      def initialize(error_message = nil, exception: nil, **)
         error = I18n.t('api_v3.errors.code_500')
 
-        if error_message
+        if error_message && !hidden_exception?(exception)
```
Comment (ulferts) [inline lib/api/errors/internal_error.rb]: If you negate the `hidden`, why not call it `shown` or `displayed`?

**Review by ulferts — CHANGES_REQUESTED** [review]

File: lib/api/errors/internal_error.rb (hunk 1 above)
Comment (oliverguenther) [reply lib/api/errors/internal_error.rb]: Good point. I suggest we always hide it instead then.

**Review by ulferts — APPROVED** [review]

**Pull request merged**